keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,539 @@
+- name: v1
+  served: true
+  schema:
+    openAPIV3Schema:
+      properties:
+        foo:
+          type: object
+          x-kubernetes-immutable: true
+          properties:
+            a:
+              type: string
+            b:
+              type: string
+- name: v2
+  served: true
+  storage: true
+  schema:
+    openAPIV3Schema:
+      properties:
+        foo:
+          type: object
+          x-kubernetes-immutable: true
+          properties:
+            a:
+              type: string
+              default: A
+            b:
+              type: string    
+```
+
+In this example, `foo.a` is defaulted in v2, but not in v1. 
+
+- Assuming there is an instance in the storage version `v2` in etcd:
+
+  ```
+  apiVersion: group/v2
+  foo:
+    b: text
+  ```
+
+  it will get `foo.a` defaulted to `A` on read.
+ 
+  When updating that object with a no-op patch, the defaulting takes place on read from etcd, converted, the patch applied and the immutability check will succeed.
+
+  When reading the object with a GET request, the `foo.a` field is defaulted and the client sending an update will have that value such that the immutability check will succeed as well.
+  
+- Now assume the storage version is `v1` and we have stored
+
+  ```
+  apiVersion: group/v1
+  foo:
+    b: text
+  ```
+
+  The `v2` GET request will read from etcd in the storage version `v1`. `v1` has no default such that `foo.a` is not set. The conversion to `v2` does not set it either. Hence, the returned `v2` object has no `foo.a`.
+  
+  Now a `v2` update is sent with the same object just received. Now the incoming request payload is defaulted via the `v2` schema and hence `foo.a` is set. The `v1` stored object is read and converted to `v2` without setting `foo.a`. The immutability check fails.
+  
+The learning from the example is that the stored version has to have "at least the defaults as any other version". With arbitrary conversions, there is no way though to mechanically check that versions are compatible with immutability. But this is not new as even without immutability, OpenAPI schemas of multi-version CRDs with conversions must be carefully crafted by the developer to work as expected from the API consumer point of view.
+
+### Publishing
+
+The `x-kubernetes-immutable` and `x-kubernetes-immutable-keys` vendor extensions are published via `/openapi/v2` as is.
+
+### Suggested marker syntax
+
+In analogy to `+required`, `+optional` we propose to add a marker to kubebuilder's
+controller-gen named `+immutable`, which covers the whole subtree of the object 
+(compare field-selection section for extension).
+
+```
+// The name can not be changed after creation.
+// +immutable
+Name string
+
+// The list of containers can not change AT ALL after creation.
+// No single field in existing containers can be changed, added or deleted,
+// no new containers can be added, no existing container can be removed.
+// +immutable
+Containers []Containers
+```
```

Comment:
Suggestion:  Recommend a marker syntax for maps and map-slices in Go.  
Possible syntax:

```
// Okay to change any way
A map[string]string

// Can't touch this.
// +immutable
B map[string]string 

// Allowed addition and deletion.
// +immutable-map-values  <--- not sure how to express this in Go.
C map[string]string 

// Allowed addition and deletion.
// +listType=map
// +listMapKey=name
// +immutable-map-values
D []SomeType

// Allowed existing value modification.
// +immutable-keys
E map[string]string 

// Allowed existing value modification.
// +listType=map
// +listMapKey=name
// +immutable-keys
F []SomeType
```
// Can't add ports to running pod but can  existing options but can add new ones.
// +immutable=values  <--- not sure how to express this in Go.
Options map[string]string

Reply:
cc @DirectXMan12

Reply:
@erictune there is a discussion (for quite some time) about how to express properties of items and map values. Am not sure this is so different here, at least it is overlapping.

I'd like to have general `+kubebuilder:validation:items:<some-marker>` and `+kubebuilder:validation:values:<some-marker>` syntax.